File: lint-arwaky/archive/capabilities/dependency_cycle_analyzer.py
"""
Cycle detection (DFS) for directed graphs.

"""

from __future__ import annotations
from typing import Any
import logging

from ..taxonomy import (
    AdapterName,
    ArchitectureConfig,
    ColumnNumber,
    ComplianceStatus,
    ContentString,
    DirectoryPath,
    ErrorCode,
    FilePath,
    Identity,
    LayerNameVO,
    LineNumber,
    LintMessage,
    LintResult,
    LintResultList,
    Location,
    LocationList,
    PatternList,
    Severity,
    SymbolName,
    LAYER_AGENT,
)
from ..contract import (
    ICycleAnalysisProtocol,
    IFileSystemPort,
    ISourceParserPort,
    ISemanticTracerProtocol,
)

logger = logging.getLogger(__name__)

try:
    from auto_linter import auto_linter_rust
    RUST_AVAILABLE = True
except ImportError:
    RUST_AVAILABLE = False


def _detect_cycle_edges(edges: list[dict[str, str]]) -> PatternList:
    """Detect cycles in a directed graph adjacency list.

    Returns list of "source->target" edge keys that are part of cycles.
    Uses DFS-based cycle detection.
    """
    if RUST_AVAILABLE:
        try:
            res_list = auto_linter_rust.detect_cycle_edges(edges)
            return PatternList(values=res_list)
        except Exception as e:
            logger.warning(f"Rust cycle detector failed: {e}. Falling back to Python.")

    from collections import defaultdict

    graph: dict[ContentString, set[ContentString]] = defaultdict(set)
    for e in edges:
        graph[e["source"]].add(e["target"])

    WHITE, GRAY, BLACK = 0, 1, 2
    color: dict[ContentString, int] = defaultdict(lambda: WHITE)
    cycle_edges: PatternList = PatternList(values=[])

    def dfs(node: ContentString, path: set[ContentString]) -> None:
        color[node] = GRAY
        for neigh in list(graph.get(node, [])):
            if color[neigh] == GRAY:
                cycle_edges.values.append(f"{node}->{neigh}")
            elif color[neigh] == WHITE:
                dfs(neigh, path)
        color[node] = BLACK

    for node in list(graph):
        if color[node] == WHITE:
            dfs(node, set())

    return cycle_edges


class CycleAnalyzer(ICycleAnalysisProtocol):
    """Detects circular imports and dependency cycles (Capability)."""

    def __init__(
        self,
        config: ArchitectureConfig,
        layer_map: dict[Identity, Any],
        fs: IFileSystemPort,
        parser: ISourceParserPort,
        tracer: ISemanticTracerProtocol | None = None,
    ):
        self.config = config
        self.fs = fs
        self.parser = parser
        self.tracer = tracer
        self.layer_map = layer_map

    def name(self) -> AdapterName:
        return AdapterName(value="architecture")

    async def scan(self, path: FilePath) -> LintResultList:
        """Scan path (file or directory) for circular import violations."""
        if not self.config.enabled or not self.config.governance_rules:
            return LintResultList()

        results: LintResultList = LintResultList(values=[])
        root_dir = self._resolve_root(path)

        # Walk all Python files
        for file_path in self.fs.walk(path):
            if not str(file_path).endswith(".py"):
                continue
            file_layer = self._detect_layer(file_path, root_dir)
            if not file_layer or file_layer == LAYER_AGENT:
                continue

            results.values.extend(self._scan_file(file_path, file_layer, root_dir))

        # Record history for trends analysis
        self._record_history(str(path), results)
        return results

    def _scan_file(
        self, file_path: FilePath, file_layer: LayerNameVO, root_dir: FilePath
    ) -> list[LintResult]:
        """Process a single Python file for circular import violations."""
        violations: list[LintResult] = []
        imports = self.parser.extract_imports(file_path)

        # Build dependency graph from imports
        edges: list[dict[str, str]] = []
        for imp in imports:
            target_layer = self._detect_module_layer(imp.module)
            if target_layer:
                edges.append({"source": str(file_layer), "target": str(target_layer)})

        # Detect cycles in the dependency graph
        cycle_edges = _detect_cycle_edges(edges)
        if cycle_edges:
            # Create violation for each cycle edge
            for edge_key in cycle_edges:
                source, target = edge_key.split("->")
                violation = self._create_violation(
                    file_path=file_path,
                    line_no=1,  # We don't have line info for graph-level violations
                    module_path=f"{source} -> {target}",
                    description="Circular import detected",
                    file_layer=str(file_layer),
                    target_layer=target,
                    root_dir=root_dir,
                    imported_name=None,
                )
                violations.append(violation)

        return violations

    def _create_violation(
        self,
        file_path: FilePath,
        line_no: LineNumber,
        module_path: ContentString,
        description: ContentString,
        file_layer: Identity,
        target_layer: Identity,
        root_dir: FilePath,
        imported_name: Identity | None = None,
    ) -> LintResult:
        """Construct a LintResult for a circular import violation."""
        base_msg: ContentString = ContentString(
            value=(
                f"[AES Circular Import] {description}. "
                f"Dependency cycle: {file_layer} -> {target_layer} via '{module_path}'."
            )
        )

        # Optional: extend with tracer information if available
        related_locations: list[ContentString] = []
        if self.tracer and imported_name and str(imported_name) != "*":
            try:
                func_candidate = str(imported_name)
                callers = self.tracer.trace_call_chain(
                    root_dir=DirectoryPath(value=str(root_dir)),
                    target_name=SymbolName(value=func_candidate),
                )
                related_locations = [
                    ContentString(value=f"CallSite: {c}") for c in callers.values[:3]
                ]
            except Exception as e:
                logger.debug(f"Tracer error: {e}")

        return LintResult(
            file=file_path,
            line=line_no,
            column=ColumnNumber(value=0),
            code=ErrorCode(code="AES020"),  # Assuming AES020 for circular imports
            message=LintMessage(value=str(base_msg)),
            source=AdapterName(value=str(self.name())),
            severity=Severity.CRITICAL,
            related_locations=LocationList(
                values=[Location(description=loc) for loc in related_locations]
            ),
        )

    def apply_fix(self, path: FilePath) -> ComplianceStatus:
        """Circular import violations require manual architectural refactoring."""
        return ComplianceStatus(value=False)

    def _resolve_root(self, path: FilePath) -> FilePath:
        """Find project root (directory containing 'src')."""
        current = path
        for _ in range(5):
            src_dir = self.fs.path_join(str(current), "src")
            if self.fs.is_directory(src_dir).value:
                return current
            parent = self.fs.get_parent(current)
            if parent == current:
                break
            current = parent
        return path

    def _detect_layer(
        self, file_path: FilePath, root_dir: FilePath
    ) -> LayerNameVO | None:
        rel: ContentString = ContentString(
            value=str(self.fs.get_relative_path(file_path, root_dir))
        )
        sorted_layers = sorted(
            self.layer_map.items(), key=lambda x: len(str(x[1].path)), reverse=True
        )
        for name, definition in sorted_layers:
            path_def = definition.path_str
            if rel.startswith(path_def) or f"/{path_def}/" in f"/{rel}/":
                return LayerNameVO(value=str(name))
        return None

    def _detect_module_layer(self, module_path: str) -> LayerNameVO | None:
        parts = str(module_path).split(".")
        for part in parts:
            for name, definition in self.layer_map.items():
                if part == name or part == definition.path_str:
                    return LayerNameVO(value=str(name))
        return None

File: lint-arwaky/src-rust/capabilities/dependency_cycle_analyzer.rs
// dependency_cycle_analyzer — Detects circular imports and dependency cycles.
// Implements ICycleAnalysisProtocol: scan files for circular import violations.

use crate::taxonomy::{
    AdapterName, ArchitectureConfig, ColumnNumber, ErrorCode, FilePath, LayerNameVO, LineNumber,
    LintMessage, LintResult, LocationList, ScopeRef, Severity, SymbolName,
};
use std::collections::{HashMap, HashSet};
use std::fs;

/// Represents a single edge in a dependency graph.
pub struct DependencyEdge {
    pub source: String,
    pub target: String,
}

impl DependencyEdge {
    pub fn new(source: String, target: String) -> Self {
        Self { source, target }
    }
}

/// Detects cycles in a directed graph adjacency list.
/// Returns list of "source->target" edge keys that are part of cycles.
pub fn detect_cycle_edges(edges: &[DependencyEdge]) -> Vec<SymbolName> {
    let mut graph: HashMap<String, HashSet<String>> = HashMap::new();
    for e in edges {
        graph
            .entry(e.source.clone())
            .or_default()
            .insert(e.target.clone());
    }

    let mut cycle_edges: Vec<String> = Vec::new();
    let mut visited: HashSet<String> = HashSet::new();
    let mut path_stack: HashSet<String> = HashSet::new();

    fn dfs(
        node: &str,
        graph: &HashMap<String, HashSet<String>>,
        visited: &mut HashSet<String>,
        path_stack: &mut HashSet<String>,
        cycle_edges: &mut Vec<String>,
    ) {
        visited.insert(node.to_string());
        path_stack.insert(node.to_string());

        if let Some(neighbors) = graph.get(node) {
            for neighbor in neighbors {
                if path_stack.contains(neighbor) {
                    cycle_edges.push(format!("{}->{}", node, neighbor));
                } else if !visited.contains(neighbor) {
                    dfs(neighbor, graph, visited, path_stack, cycle_edges);
                }
            }
        }

        path_stack.remove(node);
    }

    let nodes: Vec<String> = graph.keys().cloned().collect();
    for node in nodes {
        if !visited.contains(&node) {
            dfs(
                &node,
                &graph,
                &mut visited,
                &mut path_stack,
                &mut cycle_edges,
            );
        }
    }

    cycle_edges
        .into_iter()
        .map(|s| SymbolName::new(s))
        .collect()
}

/// Detects circular imports and dependency cycles (Capability).
pub struct DependencyCycleAnalyzer {
    config: ArchitectureConfig,
}

impl DependencyCycleAnalyzer {
    pub fn new(config: ArchitectureConfig) -> Self {
        Self { config }
    }

    fn make_result(file: &str, msg: &str) -> LintResult {
        LintResult {
            file: FilePath::new(file.to_string()),
            line: LineNumber::new(1),
            column: ColumnNumber::new(0),
            code: ErrorCode::new("AES020"),
            message: LintMessage::new(msg),
            source: AdapterName::new("architecture"),
            severity: Severity::CRITICAL,
            enclosing_scope: ScopeRef {
                name: "".to_string(),
                kind: "".to_string(),
                file: FilePath::new(""),
                start_line: LineNumber::new(0),
                end_line: LineNumber::new(0),
            },
            related_locations: LocationList::new(Vec::new()),
        }
    }

    fn extract_import_modules(content: &str) -> Vec<String> {
        let mut modules = Vec::new();
        for line in content.lines() {
            let trimmed = line.trim();
            if let Some(rest) = trimmed.strip_prefix("from ") {
                if let Some(module) = rest.split_whitespace().next() {
                    modules.push(module.to_string());
                }
            } else if let Some(rest) = trimmed.strip_prefix("import ") {
                if let Some(module) = rest.split_whitespace().next() {
                    modules.push(module.trim_end_matches(',').to_string());
                }
            } else if let Some(rest) = trimmed.strip_prefix("use ") {
                let module = rest.trim_end_matches(';');
                modules.push(module.to_string());
            }
        }
        modules
    }

    fn detect_file_layer(&self, file: &str, root_dir: &str) -> Option<String> {
        let rel = file
            .strip_prefix(root_dir)
            .unwrap_or(file)
            .trim_start_matches('/');
        let mut layers: Vec<(&LayerNameVO, &crate::taxonomy::LayerDefinition)> =
            self.config.layers.iter().collect();
        layers.sort_by(|a, b| b.1.path.value.len().cmp(&a.1.path.value.len()));

        for (name, def) in layers {
            let layer_path = def.path.value.as_str();
            if rel.starts_with(layer_path) {
                return Some(name.value.clone());
            }
        }
        None
    }

    fn detect_module_layer(&self, module: &str) -> Option<String> {
        let parts: Vec<&str> = module.split('.').collect();
        for part in &parts {
            for (name, _def) in &self.config.layers {
                if *part == name.value.as_str() {
                    return Some(name.value.clone());
                }
            }
        }
        None
    }

    pub fn scan(&self, files: &[String], root_dir: &str) -> Vec<LintResult> {
        if !self.config.enabled {
            return vec![];
        }

        let mut edges = Vec::new();
        let mut file_by_layer: HashMap<String, String> = HashMap::new();

        for file in files {
            let Ok(content) = fs::read_to_string(file) else {
                continue;
            };
            let file_layer = match self.detect_file_layer(file, root_dir) {
                Some(l) => l,
                None => continue,
            };

            file_by_layer
                .entry(file_layer.clone())
                .or_insert_with(|| file.clone());

            let modules = Self::extract_import_modules(&content);
            for module in modules {
                if let Some(target_layer) = self.detect_module_layer(&module) {
                    if target_layer != file_layer {
                        edges.push(DependencyEdge::new(file_layer.clone(), target_layer));
                    }
                }
            }
        }

        let cycle_edge_results = detect_cycle_edges(&edges);

        cycle_edge_results
            .into_iter()
            .map(|sn| {
                let edge_key = sn.value;
                let parts: Vec<&str> = edge_key.split("->").collect();
                let source = parts[0];
                let target = parts[1];
                let file = file_by_layer
                    .get(source)
                    .cloned()
                    .unwrap_or_else(|| source.to_string());
                let msg = format!(
                    "[AES Circular Import] Dependency cycle detected: '{}' -> '{}'.",
                    source, target
                );
                Self::make_result(&file, &msg)
            })
            .collect()
    }
}
